--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -745,11 +_,12 @@
             this.setMaxDamage(maxDamage);
         }
 
+        CompoundTag internalTag = null;
         String internal = SerializableMeta.getString(map, "internal", true);
         if (internal != null) {
             ByteArrayInputStream buf = new ByteArrayInputStream(Base64.getDecoder().decode(internal));
             try {
-                CompoundTag internalTag = NbtIo.readCompressed(buf, NbtAccounter.unlimitedHeap());
+                internalTag = NbtIo.readCompressed(buf, NbtAccounter.unlimitedHeap());
                 this.deserializeInternal(internalTag, map);
             } catch (IOException ex) {
                 LOGGER.error("Failed to read internal tag from object", ex);
@@ -812,6 +_,18 @@
                 this.customTag = NbtIo.readCompressed(buf, NbtAccounter.unlimitedHeap());
             } catch (IOException ex) {
                 LOGGER.error("Failed to read custom tag for item", ex);
+            }
+        }
+
+        if (internalTag != null) {
+            if (this.customTag == null) {
+                this.customTag = new CompoundTag();
+            }
+            for (String key : internalTag.keySet()) {
+                if (this.customTag.contains(key)) {
+                    continue;
+                }
+                this.customTag.put(key, internalTag.get(key));
             }
         }
     }
