--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
@@ -19,7 +_,11 @@
 
 public interface CraftRecipe extends Recipe {
 
-    void addToRecipeManager();
+    void addToRecipeManager(boolean reload);
+
+    default void addToRecipeManager() {
+        this.addToRecipeManager(true);
+    }
 
     static Optional<Ingredient> toPossibleIngredient(@Nullable RecipeChoice bukkit, boolean requireNotEmpty) {
         return (bukkit == null || bukkit == RecipeChoice.empty()) ? Optional.empty() : Optional.of(toIngredient(bukkit, requireNotEmpty)); // Paper - support "empty" choices
