--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -817,7 +_,7 @@
                     this.awardStat(Stats.CROUCH_TIME);
                 }
 
-                if (!this.isSleeping()) {
+                if (!this.isSleeping() && this.level().paperConfig().entities.behavior.tickTimeSinceSleep) {
                     this.awardStat(Stats.TIME_SINCE_REST);
                 }
             }
@@ -2683,7 +_,7 @@
                 event.isForced()
             );
             if (event.willNotifyPlayer() && event.getNotification() != null) {
-                this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getNotification()));
+                // this.sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getNotification()));
             }
         }
 
