--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1056,7 +_,7 @@
 
     private void announceSleepStatus() {
         if (this.canSleepThroughNights()) {
-            if (!this.getServer().isSingleplayer() || this.getServer().isPublished()) {
+            if (false && !this.getServer().isSingleplayer() || this.getServer().isPublished()) {
                 int _int = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
                 Component component;
                 if (this.sleepStatus.areEnoughSleeping(_int)) {
@@ -1648,34 +_,48 @@
     public void playSeededSound(
         @Nullable Entity entity, double x, double y, double z, Holder<SoundEvent> sound, SoundSource source, float volume, float pitch, long seed
     ) {
-        this.server
-            .getPlayerList()
-            .broadcast(
-                entity instanceof Player player ? player : null,
-                x,
-                y,
-                z,
-                sound.value().getRange(volume),
-                this.dimension(),
-                new ClientboundSoundPacket(sound, source, x, y, z, volume, pitch, seed)
-            );
+        // Parchment start - sound event
+        org.bukkit.craftbukkit.event.CraftEventFactory.playSoundEvent(new gg.projecteden.parchment.event.sound.SoundEvent(
+                entity == null ? null : entity.getBukkitEntity(),
+            net.kyori.adventure.sound.Sound.sound()
+                    .type(sound.unwrap().<net.kyori.adventure.key.Key>map(
+                        key -> io.papermc.paper.adventure.PaperAdventure.asAdventure(key.location()),
+                        soundEvent -> io.papermc.paper.adventure.PaperAdventure.asAdventure(soundEvent.location())
+                    ))
+                    .source(io.papermc.paper.adventure.PaperAdventure.asAdventure(source))
+                    .volume(volume)
+                    .pitch(pitch)
+                    .seed(seed)
+                    .build(),
+            gg.projecteden.parchment.event.sound.ParchmentSoundEvent.createEmitter(this, x, y, z),
+            gg.projecteden.parchment.event.sound.ParchmentSoundEvent.DISTANCE_FUNCTION,
+            null
+        ));
+        // Parchment end
     }
 
     @Override
     public void playSeededSound(
         @Nullable Entity entity, Entity sourceEntity, Holder<SoundEvent> sound, SoundSource source, float volume, float pitch, long seed
     ) {
-        this.server
-            .getPlayerList()
-            .broadcast(
-                entity instanceof Player player ? player : null,
-                sourceEntity.getX(),
-                sourceEntity.getY(),
-                sourceEntity.getZ(),
-                sound.value().getRange(volume),
-                this.dimension(),
-                new ClientboundSoundEntityPacket(sound, source, sourceEntity, volume, pitch, seed)
-            );
+        // Parchment start - sound event
+        org.bukkit.craftbukkit.event.CraftEventFactory.playSoundEvent(new gg.projecteden.parchment.event.sound.SoundEvent(
+                entity == null ? null : entity.getBukkitEntity(),
+            net.kyori.adventure.sound.Sound.sound()
+                .type(sound.unwrap().<net.kyori.adventure.key.Key>map(
+                    key -> io.papermc.paper.adventure.PaperAdventure.asAdventure(key.location()),
+                    soundEvent -> io.papermc.paper.adventure.PaperAdventure.asAdventure(soundEvent.location())
+                ))
+                .source(io.papermc.paper.adventure.PaperAdventure.asAdventure(source))
+                .volume(volume)
+                .pitch(pitch)
+                .seed(seed)
+                .build(),
+            gg.projecteden.parchment.event.sound.ParchmentSoundEvent.createEmitter(sourceEntity),
+            gg.projecteden.parchment.event.sound.ParchmentSoundEvent.DISTANCE_FUNCTION,
+            null
+        ));
+        // Parchment end
     }
 
     @Override
