--- a/src/main/java/io/papermc/paper/event/player/PlayerPickItemEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/PlayerPickItemEvent.java
@@ -1,6 +_,8 @@
 package io.papermc.paper.event.player;
 
 import com.google.common.base.Preconditions;
+import org.bukkit.Location;
+import org.bukkit.entity.Entity;
 import org.bukkit.entity.Player;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
@@ -24,6 +_,8 @@
 
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
+    private final Location location;
+    private final Entity entity;
     private final boolean includeData;
 
     private int targetSlot;
@@ -32,11 +_,13 @@
     private boolean cancelled;
 
     @ApiStatus.Internal
-    protected PlayerPickItemEvent(final Player player, final boolean includeData, final int targetSlot, final int sourceSlot) {
+    public PlayerPickItemEvent(final Player player, final int targetSlot, final int sourceSlot, final Location location, final Entity entity, boolean includeData) {
         super(player);
         this.includeData = includeData;
         this.targetSlot = targetSlot;
         this.sourceSlot = sourceSlot;
+        this.location = location;
+        this.entity = entity;
     }
 
     /**
@@ -89,6 +_,14 @@
     public void setSourceSlot(final @Range(from = -1, to = 35) int sourceSlot) {
         Preconditions.checkArgument(sourceSlot >= -1 && sourceSlot <= 35, "Source slot must be in range of the player's inventory slot, or -1");
         this.sourceSlot = sourceSlot;
+    }
+
+    public @org.jetbrains.annotations.Nullable Location getLocation() {
+        return location;
+    }
+
+    public @org.jetbrains.annotations.Nullable Entity getEntity() {
+        return entity;
     }
 
     @Override
