--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -26,6 +_,8 @@
     private static final EntityDataAccessor<Direction> DATA_DIRECTION = SynchedEntityData.defineId(HangingEntity.class, EntityDataSerializers.DIRECTION);
     private static final Direction DEFAULT_DIRECTION = Direction.SOUTH;
 
+    public boolean tick = true; // Parchment
+
     protected HangingEntity(final EntityType<? extends HangingEntity> type, final Level level) {
         super(type, level);
     }
@@ -161,4 +_,12 @@
     public float mirror(final Mirror mirror) {
         return this.rotate(mirror.getRotation(this.getDirection()));
     }
+
+    @Override
+    public void tick() {
+        if (tick) {
+            super.tick();
+        }
+    }
+
 }
