--- a/src/main/java/io/papermc/paper/adventure/PaperAdventure.java
+++ b/src/main/java/io/papermc/paper/adventure/PaperAdventure.java
@@ -435,4 +_,37 @@
             .parse(ops, encoded).getOrThrow(IllegalStateException::new);
     }
 
+
+    // Parchment start
+    public static net.kyori.adventure.sound.Sound.Source asAdventure(final SoundSource source) {
+        return switch (source) {
+            case MASTER -> net.kyori.adventure.sound.Sound.Source.MASTER;
+            case MUSIC -> net.kyori.adventure.sound.Sound.Source.MUSIC;
+            case RECORDS -> net.kyori.adventure.sound.Sound.Source.RECORD;
+            case WEATHER -> net.kyori.adventure.sound.Sound.Source.WEATHER;
+            case BLOCKS -> net.kyori.adventure.sound.Sound.Source.BLOCK;
+            case HOSTILE -> net.kyori.adventure.sound.Sound.Source.HOSTILE;
+            case NEUTRAL -> net.kyori.adventure.sound.Sound.Source.NEUTRAL;
+            case PLAYERS -> net.kyori.adventure.sound.Sound.Source.PLAYER;
+            case AMBIENT -> net.kyori.adventure.sound.Sound.Source.AMBIENT;
+            case VOICE -> net.kyori.adventure.sound.Sound.Source.VOICE;
+            case UI -> net.kyori.adventure.sound.Sound.Source.UI;
+        };
+    }
+
+    public static Optional<SoundEvent> asVanillaSound(final Key key) {
+        return BuiltInRegistries.SOUND_EVENT.getOptional(asVanilla(key));
+    }
+
+    public static Optional<net.minecraft.core.Holder.Reference<SoundEvent>> asSoundHolder(final ResourceKey<SoundEvent> key) {
+        return BuiltInRegistries.SOUND_EVENT.get(key);
+    }
+
+    public static Optional<net.minecraft.core.Holder.Reference<SoundEvent>> asVanillaSoundHolder(final Key key) {
+        return asSoundHolder(ResourceKey.create(net.minecraft.core.registries.Registries.SOUND_EVENT, asVanilla(key)));
+    }
+
+    // Parchment end
+
+
 }
