--- a/src/main/java/com/destroystokyo/paper/event/server/AsyncTabCompleteEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/server/AsyncTabCompleteEvent.java
@@ -53,7 +_,7 @@
  * Only 1 process will be allowed to provide completions, the Async Event, or the standard process.
  */
 @NullMarked
-public class AsyncTabCompleteEvent extends Event implements Cancellable {
+public class AsyncTabCompleteEvent extends Event implements Cancellable, gg.projecteden.parchment.OptionalLocation { // Parchment
 
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
