--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3111,7 +_,7 @@
         if (!this.level().isClientSide()
             && player.isSecondaryUseActive()
             && this instanceof Leashable leashable
-            && leashable.canBeLeashed()
+            && new gg.projecteden.parchment.event.entity.CheckEntityLeashableEvent(this.getBukkitEntity(), leashable.canBeLeashed()).callEvent()
             && this.isAlive()
             && !(this instanceof LivingEntity le && le.isBaby())) {
             List<Leashable> mobsToLeash = Leashable.leashableInArea(this, l -> l.getLeashHolder() == player);
@@ -5456,4 +_,30 @@
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end
+
+    @javax.annotation.Nullable
+    private gg.projecteden.parchment.entity.EntityData storedEntityData;
+
+    /**
+     * Retrieves the stored EntityData for this entity
+     * @return The currently stored EntityData
+     */
+    public gg.projecteden.parchment.entity.EntityData getStoredEntityData() {
+        if (this.storedEntityData == null) {
+            this.storedEntityData = gg.projecteden.parchment.entity.EntityData.create(this.getBukkitEntity());
+        }
+        return this.storedEntityData;
+    }
+
+    /**
+     * Clears the currently stored EntityData for this entity
+     * @return the previously stored EntityData
+     */
+    public @javax.annotation.Nullable gg.projecteden.parchment.entity.EntityData clearStoredEntityData() {
+        gg.projecteden.parchment.entity.EntityData data = this.storedEntityData;
+        this.storedEntityData = null;
+
+        return data;
+    }
+
 }
