--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5364,4 +_,30 @@
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    @javax.annotation.Nullable
+    private gg.projecteden.parchment.entity.EntityData storedEntityData;
+
+    /**
+     * Retrieves the stored EntityData for this entity
+     * @return The currently stored EntityData
+     */
+    public gg.projecteden.parchment.entity.EntityData getStoredEntityData() {
+        if (this.storedEntityData == null) {
+            this.storedEntityData = gg.projecteden.parchment.entity.EntityData.create(this.getBukkitEntity());
+        }
+        return this.storedEntityData;
+    }
+
+    /**
+     * Clears the currently stored EntityData for this entity
+     * @return the previously stored EntityData
+     */
+    public @javax.annotation.Nullable gg.projecteden.parchment.entity.EntityData clearStoredEntityData() {
+        gg.projecteden.parchment.entity.EntityData data = this.storedEntityData;
+        this.storedEntityData = null;
+
+        return data;
+    }
+
 }
