--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -54,7 +_,7 @@
     }
 
     default boolean canHaveALeashAttachedTo(final Entity entity) {
-        return this != entity && !(this.leashDistanceTo(entity) > this.leashSnapDistance()) && this.canBeLeashed();
+        return this != entity && !(this.leashDistanceTo(entity) > this.leashSnapDistance()) && new gg.projecteden.parchment.event.entity.CheckEntityLeashableEvent(((Entity) this).getBukkitEntity(), this.canBeLeashed()).callEvent();
     }
 
     default double leashDistanceTo(final Entity entity) {
