--- a/net/minecraft/util/datafix/fixes/UnflattenTextComponentFix.java
+++ b/net/minecraft/util/datafix/fixes/UnflattenTextComponentFix.java
@@ -44,6 +_,9 @@
                 return new Dynamic<>(ops, JsonOps.INSTANCE.convertTo(ops, json));
             }
         } catch (Exception e) {
+            if (!jsonString.startsWith("\"") && !jsonString.endsWith("\"")) {
+                return unflattenJson(ops, "\"" + jsonString + "\"");
+            }
             LOGGER.error("Failed to unflatten text component json: {}", jsonString, e);
         }
 
