--- a/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -51,6 +_,7 @@
         boolean isCrit,
         @Nullable LivingEntity target
         ,float drawStrength // Paper - Pass draw strength
+        ,boolean relative
     ) {
         float f = EnchantmentHelper.processProjectileSpread(level, weapon, shooter, 0.0F);
         float f1 = projectileItems.size() == 1 ? 0.0F : 2.0F * f / (projectileItems.size() - 1);
@@ -65,7 +_,7 @@
                 int i1 = i;
                 // CraftBukkit start
                 Projectile projectile = this.createProjectile(level, shooter, weapon, itemStack, isCrit);
-                this.shootProjectile(shooter, projectile, i1, velocity, inaccuracy, f4, target);
+                this.shootProjectile(shooter, projectile, i1, velocity, inaccuracy, f4, target, relative);
 
                 org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(shooter, weapon, itemStack, projectile, hand, drawStrength, true);
                 if (event.isCancelled()) {
@@ -96,7 +_,7 @@
     }
 
     protected abstract void shootProjectile(
-        LivingEntity shooter, Projectile projectile, int index, float velocity, float inaccuracy, float angle, @Nullable LivingEntity target
+        LivingEntity shooter, Projectile projectile, int index, float velocity, float inaccuracy, float angle, @Nullable LivingEntity target, boolean relative
     );
 
     protected Projectile createProjectile(Level level, LivingEntity shooter, ItemStack weapon, ItemStack ammo, boolean isCrit) {
