--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -455,11 +_,20 @@
 
     @Override
     public boolean breakNaturally(ItemStack item, boolean triggerEffect, boolean dropExperience) {
-        return this.breakNaturally(item, triggerEffect, dropExperience, false);
+        return this.breakNaturally(null, item, triggerEffect, dropExperience, false);
     }
 
     @Override
     public boolean breakNaturally(ItemStack item, boolean triggerEffect, boolean dropExperience, boolean forceEffect) {
+        return this.breakNaturally(null, item, triggerEffect, dropExperience, forceEffect);
+    }
+
+    @Override
+    public boolean breakNaturally(Player player, ItemStack item, boolean triggerEffect, boolean dropExperience) {
+        return this.breakNaturally(null, item, triggerEffect, dropExperience, false);
+    }
+
+    public boolean breakNaturally(Player player, ItemStack item, boolean triggerEffect, boolean dropExperience, boolean forceEffect) {
         // Order matters here, need to drop before setting to air so skulls can get their data
         BlockState state = this.getBlockState();
         net.minecraft.world.level.block.Block block = state.getBlock();
@@ -468,7 +_,11 @@
 
         // Modelled off Player#hasCorrectToolForDrops
         if (!state.isAir() && (item == null || !state.requiresCorrectToolForDrops() || nmsItem.isCorrectToolForDrops(state))) {
-            net.minecraft.world.level.block.Block.dropResources(state, this.level.getMinecraftWorld(), this.position, this.level.getBlockEntity(this.position), null, nmsItem, false);
+            List<net.minecraft.world.entity.item.ItemEntity> itemEntities = net.minecraft.world.level.block.Block.dropResources(state, this.level.getMinecraftWorld(), this.position, this.level.getBlockEntity(this.position), null, nmsItem, false);
+
+            if (player != null) {
+                new org.bukkit.event.block.BlockDropItemEvent(this, this.getState(), player, itemEntities.stream().map(i -> (org.bukkit.entity.Item) CraftEntity.getEntity((org.bukkit.craftbukkit.CraftServer) Bukkit.getServer(), i)).toList()).callEvent();
+            }
             if (dropExperience) block.popExperience(this.level.getMinecraftWorld(), this.position, block.getExpDrop(state, this.level.getMinecraftWorld(), this.position, nmsItem, true));
             result = true;
         }
