--- a/src/main/java/org/bukkit/event/raid/RaidTriggerEvent.java
+++ b/src/main/java/org/bukkit/event/raid/RaidTriggerEvent.java
@@ -11,7 +_,7 @@
  * Called when a {@link Raid} is triggered (e.g: a player with Bad Omen effect
  * enters a village).
  */
-public class RaidTriggerEvent extends RaidEvent implements Cancellable {
+public class RaidTriggerEvent extends RaidEvent implements Cancellable, gg.projecteden.parchment.HasPlayer { // Parchment
 
     private static final HandlerList handlers = new HandlerList();
     //
