--- a/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
+++ b/src/main/java/org/bukkit/event/block/BlockDropItemEvent.java
@@ -28,7 +_,7 @@
  * AIR in most cases. Use #getBlockState() for more Information about the broken
  * block.
  */
-public class BlockDropItemEvent extends BlockEvent implements Cancellable {
+public class BlockDropItemEvent extends BlockEvent implements Cancellable, gg.projecteden.parchment.HasPlayer { // Parchment
 
     private static final HandlerList handlers = new HandlerList();
     private final Player player;
