--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -767,19 +_,7 @@
                 BlockPos blockPos = respawnConfig.pos();
                 BlockState blockState = level1.getBlockState(blockPos);
                 if (blockState.is(Blocks.RESPAWN_ANCHOR)) {
-                    serverPlayer.connection
-                        .send(
-                            new ClientboundSoundPacket(
-                                SoundEvents.RESPAWN_ANCHOR_DEPLETE,
-                                SoundSource.BLOCKS,
-                                blockPos.getX(),
-                                blockPos.getY(),
-                                blockPos.getZ(),
-                                1.0F,
-                                1.0F,
-                                level.getRandom().nextLong()
-                            )
-                        );
+                    player.playNotifySound(SoundEvents.RESPAWN_ANCHOR_DEPLETE.value(), SoundSource.BLOCKS, 1.0F, 1.0F); // Parchment - use existing play sound method
                 }
                 // Paper start - Add PlayerPostRespawnEvent
                 if (!teleportTransition.missingRespawnBlock()) {
