--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -663,19 +_,23 @@
                 BlockPos respawnPosition = respawnData.pos();
                 BlockState blockState = respawnLevel.getBlockState(respawnPosition);
                 if (blockState.is(Blocks.RESPAWN_ANCHOR)) {
-                    player.connection
-                        .send(
-                            new ClientboundSoundPacket(
-                                SoundEvents.RESPAWN_ANCHOR_DEPLETE,
-                                SoundSource.BLOCKS,
-                                respawnPosition.getX(),
-                                respawnPosition.getY(),
-                                respawnPosition.getZ(),
-                                1.0F,
-                                1.0F,
-                                level.getRandom().nextLong()
-                            )
-                        );
+
+                    org.bukkit.craftbukkit.event.CraftEventFactory.playSoundEvent(new gg.projecteden.parchment.event.sound.SoundEvent(
+                            null,
+                            net.kyori.adventure.sound.Sound.sound()
+                                    .type(SoundEvents.RESPAWN_ANCHOR_DEPLETE.unwrap().<net.kyori.adventure.key.Key>map(
+                                            key -> io.papermc.paper.adventure.PaperAdventure.asAdventure(key.identifier()),
+                                            soundEvent -> io.papermc.paper.adventure.PaperAdventure.asAdventure(soundEvent.location())
+                                    ))
+                                    .source(io.papermc.paper.adventure.PaperAdventure.asAdventure(SoundSource.BLOCKS))
+                                    .volume(1.0f)
+                                    .pitch(1.0f)
+                                    .seed(level.getRandom().nextLong())
+                                    .build(),
+                            gg.projecteden.parchment.event.sound.ParchmentSoundEvent.createEmitter(respawnLevel, respawnPosition.getX(), respawnPosition.getY(), respawnPosition.getZ()),
+                            gg.projecteden.parchment.event.sound.ParchmentSoundEvent.DISTANCE_FUNCTION,
+                            null
+                    ));
                 }
             }
         }
