--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -161,7 +_,9 @@
         if (!sourceMovement.isFinite()) {
             sourceMovement = Vec3.ZERO;
         }
-        if (!source.level().paperConfig().misc.disableRelativeProjectileVelocity) {
+        gg.projecteden.parchment.event.entity.PreEntityShootBowEvent preEntityShootBowEvent = new gg.projecteden.parchment.event.entity.PreEntityShootBowEvent(source.getBukkitEntity());
+        preEntityShootBowEvent.callEvent();
+        if (!source.level().paperConfig().misc.disableRelativeProjectileVelocity && preEntityShootBowEvent.isRelative()) {
         this.setDeltaMovement(this.getDeltaMovement().add(sourceMovement.x, source.onGround() ? 0.0 : sourceMovement.y, sourceMovement.z));
         }
         // Paper end - allow disabling relative velocity
