--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -226,6 +_,7 @@
     public org.bukkit.event.player.PlayerResourcePackStatusEvent.Status resourcePackStatus; // Paper - more resource pack API
     private static final boolean DISABLE_CHANNEL_LIMIT = System.getProperty("paper.disableChannelLimit") != null; // Paper - add a flag to disable the channel limit
     private long lastSaveTime; // Paper - getLastPlayed replacement API
+    private boolean bypassesInsomnia = false; // Parchment - Insomnia api
 
     public CraftPlayer(CraftServer server, ServerPlayer entity) {
         super(server, entity);
@@ -3583,4 +_,39 @@
         handle.containerMenu.broadcastChanges();
         return new PaperPlayerGiveResult(leftovers.build(), drops.build());
     }
+
+    // Parchment start
+    @Override
+    public Set<UUID> getHiddenEntities(Plugin plugin) {
+        return invertedVisibilityEntities.entrySet().stream()
+            .filter(entry -> entry.getValue().contains(CraftPlayer.getPluginWeakReference(plugin)))
+            .map(Map.Entry::getKey)
+            .collect(java.util.stream.Collectors.toSet());
+    }
+
+    @Override
+    public boolean isInsomniac() {
+        return net.minecraft.world.entity.EntitySelector.IS_INSOMNIAC.test(this.getHandle());
+    }
+
+    @Override
+    public void setBypassInsomnia(boolean val) {
+        this.bypassesInsomnia = val;
+    }
+
+    @Override
+    public boolean doesBypassInsomnia() {
+        return this.bypassesInsomnia;
+    }
+
+    @Override
+    public void setTimeSinceLastRest(int ticks) {
+        this.getHandle().getStats().setValue(this.getHandle(), net.minecraft.stats.Stats.CUSTOM.get(net.minecraft.stats.Stats.TIME_SINCE_REST), net.minecraft.util.Mth.clamp(ticks, 1, Integer.MAX_VALUE));
+    }
+
+    @Override
+    public int getTimeSinceLastRest() {
+        return net.minecraft.util.Mth.clamp(this.getHandle().getStats().getValue(net.minecraft.stats.Stats.CUSTOM.get(net.minecraft.stats.Stats.TIME_SINCE_REST)), 1, Integer.MAX_VALUE);
+    }
+
 }
