--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -539,7 +_,21 @@
             double x3 = playerLoc.x + 13.0 / distBtwn * (raidLoc.x - playerLoc.x);
             double z3 = playerLoc.z + 13.0 / distBtwn * (raidLoc.z - playerLoc.z);
             if (distBtwn <= 64.0 || playersInRaid.contains(player)) {
-                player.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, x3, player.getY(), z3, 64.0F, 1.0F, seed));
+                org.bukkit.craftbukkit.event.CraftEventFactory.playSoundEvent(new gg.projecteden.parchment.event.sound.SoundEvent(
+                        null,
+                        net.kyori.adventure.sound.Sound.sound()
+                                .type(SoundEvents.RAID_HORN.unwrap().<net.kyori.adventure.key.Key>map(
+                                        key -> io.papermc.paper.adventure.PaperAdventure.asAdventure(key.identifier()),
+                                        soundEvent -> io.papermc.paper.adventure.PaperAdventure.asAdventure(soundEvent.location())
+                                ))
+                                .source(io.papermc.paper.adventure.PaperAdventure.asAdventure(SoundSource.NEUTRAL))
+                                .volume(64.0f)
+                                .pitch(1.0f)
+                                .build(),
+                        gg.projecteden.parchment.event.sound.ParchmentSoundEvent.createEmitter(level, soundOrigin.getX(), soundOrigin.getY(), soundOrigin.getZ()),
+                        gg.projecteden.parchment.event.sound.ParchmentSoundEvent.DISTANCE_FUNCTION,
+                        null
+                ));
             }
         }
     }
