--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -531,8 +_,21 @@
             double d = vec3.x + 13.0 / squareRoot * (vec31.x - vec3.x);
             double d1 = vec3.z + 13.0 / squareRoot * (vec31.z - vec3.z);
             if (squareRoot <= 64.0 || players.contains(serverPlayer)) {
-                serverPlayer.connection
-                    .send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, d, serverPlayer.getY(), d1, 64.0F, 1.0F, randomLong));
+                org.bukkit.craftbukkit.event.CraftEventFactory.playSoundEvent(new gg.projecteden.parchment.event.sound.SoundEvent(
+                        null,
+                        net.kyori.adventure.sound.Sound.sound()
+                                .type(SoundEvents.RAID_HORN.unwrap().<net.kyori.adventure.key.Key>map(
+                                        key -> io.papermc.paper.adventure.PaperAdventure.asAdventure(key.identifier()),
+                                        soundEvent -> io.papermc.paper.adventure.PaperAdventure.asAdventure(soundEvent.location())
+                                ))
+                                .source(io.papermc.paper.adventure.PaperAdventure.asAdventure(SoundSource.NEUTRAL))
+                                .volume(64.0f)
+                                .pitch(1.0f)
+                                .build(),
+                        gg.projecteden.parchment.event.sound.ParchmentSoundEvent.createEmitter(level, pos.getX(), pos.getY(), pos.getZ()),
+                        gg.projecteden.parchment.event.sound.ParchmentSoundEvent.DISTANCE_FUNCTION,
+                        null
+                ));
             }
         }
     }
