--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -510,8 +_,7 @@
             double d = vec3.x + 13.0 / squareRoot * (vec31.x - vec3.x);
             double d1 = vec3.z + 13.0 / squareRoot * (vec31.z - vec3.z);
             if (squareRoot <= 64.0 || players.contains(serverPlayer)) {
-                serverPlayer.connection
-                    .send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, d, serverPlayer.getY(), d1, 64.0F, 1.0F, randomLong));
+                serverPlayer.playNotifySound(SoundEvents.RAID_HORN.value(), SoundSource.NEUTRAL, 64.0F, 1.0F); // Parchment - use existing play sound method
             }
         }
     }
