--- a/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
@@ -22,7 +_,7 @@
  * {@link Event#isAsynchronous()}
  */
 @NullMarked
-public class PreLookupProfileEvent extends Event {
+public class PreLookupProfileEvent extends Event implements gg.projecteden.api.interfaces.OptionalUniqueId { // Parchment
 
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
@@ -52,6 +_,10 @@
      * @return The UUID of the profile if it has already been provided by a plugin
      */
     public @Nullable UUID getUUID() {
+        return this.uuid;
+    }
+
+    public UUID getUniqueId() {
         return this.uuid;
     }
 
