From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Fri, 2 Jul 2021 01:33:12 -0400
Subject: [PATCH] Build changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 7ad3e5153718f6d4ce8293a9790bc3c1158aeb8e..14caef3d68aab390de482e9ea33a91641492279b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -29,6 +29,7 @@ dependencies {
     api("org.ow2.asm:asm:9.0")
     api("org.ow2.asm:asm-commons:9.0")
     api("org.apache.logging.log4j:log4j-api:2.14.1") // Paper
+    api(project(":Parchment-Common")) // Parchment
 
     compileOnly("org.apache.maven:maven-resolver-provider:3.8.1")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.0")
