From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 26 Dec 2021 20:48:45 -0500
Subject: [PATCH] Build changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 001c2b963205012f340db0d539e4033c748124ce..ec41f7da61db5209874a9f01b1b5c3a90f1ea3ea 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -22,6 +22,7 @@ configurations.api {
 }
 
 dependencies {
+    compileOnly("gg.projecteden:eden-common:2.0.5-SNAPSHOT") // Parchment
     // api dependencies are listed transitively to API consumers
     api("commons-lang:commons-lang:2.6")
     api("com.google.guava:guava:31.0.1-jre")
