From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi Larkin <lexi@qixils.dev>
Date: Sun, 6 Aug 2023 22:01:19 -0400
Subject: [PATCH] Build changes


diff --git a/build.gradle.kts b/build.gradle.kts
index fd39ed209b20c927054b8482c400beeeeab460a3..f7786d4034afae1926bb249e282f3f204e09ba1a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -14,6 +14,9 @@ val bungeeCordChatVersion = "1.20-R0.2"
 val adventureVersion = "4.17.0"
 val slf4jVersion = "2.0.9"
 val log4jVersion = "2.17.1"
+
+val edenVersion: String by project // Parchment
+
 val apiAndDocs: Configuration by configurations.creating {
     attributes {
         attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.DOCUMENTATION))
@@ -27,6 +30,7 @@ configurations.api {
 }
 
 dependencies {
+    api("gg.projecteden:eden-interfaces:$edenVersion") // Parchment
     api("com.mojang:brigadier:1.2.9") // Paper - Brigadier command api
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:32.1.2-jre")
