From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 2 May 2021 02:53:02 -0400
Subject: [PATCH] Move UUID interfaces to eden-api


diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerConnectionCloseEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerConnectionCloseEvent.java
index 81dc840856825c6ee58a46cfda04936a529ca9af..2fe9b641dc3d319b449bc3c6da27882b8af381b8 100644
--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerConnectionCloseEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerConnectionCloseEvent.java
@@ -1,6 +1,6 @@
 package com.destroystokyo.paper.event.player;
 
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import org.bukkit.event.Event;
 import org.bukkit.event.HandlerList;
 
diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java
index d9fe6d298cdbd4355775fc35f32eb5be6253a6fb..9d2f12653782a3077b506469f83647ced84eb803 100644
--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java
@@ -1,6 +1,6 @@
 package com.destroystokyo.paper.event.player;
 
-import me.lexikiq.OptionalUniqueId;
+import eden.interfaces.OptionalUniqueId;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
 import org.apache.commons.lang.Validate;
diff --git a/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java b/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
index 5218d184b9ea12a6cac3643d897dad1eafea2cb9..5834c167439fa4404e04330cad0c1c979aec44aa 100644
--- a/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
@@ -3,7 +3,7 @@ package com.destroystokyo.paper.event.profile;
 import com.destroystokyo.paper.profile.PlayerProfile;
 import com.destroystokyo.paper.profile.ProfileProperty;
 import com.google.common.collect.ArrayListMultimap;
-import me.lexikiq.OptionalUniqueId;
+import eden.interfaces.OptionalUniqueId;
 import org.bukkit.Bukkit;
 import org.bukkit.event.Event;
 import org.bukkit.event.HandlerList;
diff --git a/src/main/java/me/lexikiq/HasUniqueId.java b/src/main/java/me/lexikiq/HasUniqueId.java
deleted file mode 100644
index 4fe1da41b9417d82f52432c3bd5b4e8b4c73f09b..0000000000000000000000000000000000000000
--- a/src/main/java/me/lexikiq/HasUniqueId.java
+++ /dev/null
@@ -1,17 +0,0 @@
-package me.lexikiq;
-
-import org.jetbrains.annotations.NotNull;
-
-import java.util.UUID;
-
-/**
- * Represents an object that has a {@link UUID}
- */
-public interface HasUniqueId extends OptionalUniqueId {
-	/**
-	 * Returns a unique and persistent id for this object
-	 *
-	 * @return unique id
-	 */
-	@NotNull UUID getUniqueId();
-}
diff --git a/src/main/java/me/lexikiq/OptionalUniqueId.java b/src/main/java/me/lexikiq/OptionalUniqueId.java
deleted file mode 100644
index da4a0422dd4f072dc15750a9e133262a344ef78a..0000000000000000000000000000000000000000
--- a/src/main/java/me/lexikiq/OptionalUniqueId.java
+++ /dev/null
@@ -1,17 +0,0 @@
-package me.lexikiq;
-
-import org.jetbrains.annotations.Nullable;
-
-import java.util.UUID;
-
-/**
- * Represents an object that may have a {@link UUID}
- */
-public interface OptionalUniqueId {
-	/**
-	 * Returns a unique and persistent id for this object which may be null
-	 *
-	 * @return unique id or null
-	 */
-	@Nullable UUID getUniqueId();
-}
diff --git a/src/main/java/org/bukkit/OfflinePlayer.java b/src/main/java/org/bukkit/OfflinePlayer.java
index e2b4f86fc3825a77a5a0c1c29b428308eee54b16..523569447a91dd71e6c8648f7e848f1387ae4a8b 100644
--- a/src/main/java/org/bukkit/OfflinePlayer.java
+++ b/src/main/java/org/bukkit/OfflinePlayer.java
@@ -3,7 +3,7 @@ package org.bukkit;
 import java.util.UUID;
 
 import me.lexikiq.HasOfflinePlayer;
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import me.lexikiq.OptionalPlayer;
 import org.bukkit.configuration.serialization.ConfigurationSerializable;
 import org.bukkit.entity.AnimalTamer;
diff --git a/src/main/java/org/bukkit/attribute/AttributeModifier.java b/src/main/java/org/bukkit/attribute/AttributeModifier.java
index b2f6e641125fb4c3389cbda8a62d6198bbcacc0f..42fa2a783af28a726d2d54352d8babff386f102b 100644
--- a/src/main/java/org/bukkit/attribute/AttributeModifier.java
+++ b/src/main/java/org/bukkit/attribute/AttributeModifier.java
@@ -5,7 +5,7 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.UUID;
 
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import org.apache.commons.lang.Validate;
 import org.bukkit.configuration.serialization.ConfigurationSerializable;
 import org.bukkit.inventory.EquipmentSlot;
diff --git a/src/main/java/org/bukkit/entity/AnimalTamer.java b/src/main/java/org/bukkit/entity/AnimalTamer.java
index b412cca0cdcb7ffa295a77cb96546d3c51c683d1..7449624e517051386af53071af1cd0abe97280de 100644
--- a/src/main/java/org/bukkit/entity/AnimalTamer.java
+++ b/src/main/java/org/bukkit/entity/AnimalTamer.java
@@ -1,6 +1,6 @@
 package org.bukkit.entity;
 
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import org.jetbrains.annotations.Nullable;
 
 public interface AnimalTamer extends HasUniqueId { // Parchment
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index b33b9bb12e16f1e491912a3fdae6db552a4f8bfa..53db0a283a84d062b074b68d5d4afd14ab8ad71d 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -3,7 +3,7 @@ package org.bukkit.entity;
 import java.util.List;
 import java.util.Set;
 
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import org.bukkit.Chunk; // Paper
 import org.bukkit.EntityEffect;
 import org.bukkit.Location;
diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
index 35e8c680309f0a67e941b749240f902be07c22e6..73f974d1c5d3dc8cfbf6612ed73764e253bbd182 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
@@ -4,7 +4,7 @@ import java.net.InetAddress;
 import java.util.UUID;
 
 import com.destroystokyo.paper.profile.PlayerProfile;
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import org.bukkit.Bukkit;
 import org.bukkit.event.Event;
 import org.bukkit.event.HandlerList;
diff --git a/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
index 90cbfe9c3bd1c6fb8da85e0fb6b24d3ad23282ce..6ef9e56a74e69c8ffdc46b6e6194d766912e116b 100644
--- a/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
@@ -3,7 +3,7 @@ package org.bukkit.event.player;
 import java.net.InetAddress;
 import java.util.UUID;
 
-import me.lexikiq.HasUniqueId;
+import eden.interfaces.HasUniqueId;
 import org.bukkit.Warning;
 import org.bukkit.event.Event;
 import org.bukkit.event.HandlerList;
