From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Fri, 18 Jun 2021 01:47:40 -0400
Subject: [PATCH] Adventure fixes


diff --git a/src/main/java/org/bukkit/Keyed.java b/src/main/java/org/bukkit/Keyed.java
index 32c92621c2c15eec14c50965f5ecda00c46e6c80..3876a03ab9df9a0aef7a5c05a38b39b07a6dd285 100644
--- a/src/main/java/org/bukkit/Keyed.java
+++ b/src/main/java/org/bukkit/Keyed.java
@@ -5,7 +5,7 @@ import org.jetbrains.annotations.NotNull;
 /**
  * Represents an object which has a {@link NamespacedKey} attached to it.
  */
-public interface Keyed {
+public interface Keyed extends net.kyori.adventure.key.Keyed { // Parchment
 
     /**
      * Return the namespaced identifier for this object.
@@ -14,4 +14,11 @@ public interface Keyed {
      */
     @NotNull
     NamespacedKey getKey();
+
+    // Parchment start
+    @Override
+    default net.kyori.adventure.key.@org.checkerframework.checker.nullness.qual.NonNull Key key() {
+        return getKey();
+    }
+    // Parchment end
 }
