From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Fri, 18 Jun 2021 03:18:47 -0400
Subject: [PATCH] Add UnsafeValues#canPlaceItemOn


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 4fcafddf3792b66c618f91e04d102f374de565a8..716d5371c0ceb5b97c7008a6ac0c7ddf58b01a9f 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -243,4 +243,17 @@ public interface UnsafeValues {
      */
     boolean isCollidable(@org.jetbrains.annotations.NotNull Material material);
     // Paper end
+
+    // Parchment start
+    /**
+     * Determines if this item can be placed on a block. Must be run synchronously.
+     * Factors in player location, blocking events, hitboxes, and more.
+     * @param item item to test
+     * @param player optional player placing the block
+     * @param block block that is being placed against
+     * @param face cardinal direction
+     * @return if the item can be placed
+     */
+    java.util.concurrent.CompletableFuture<Boolean> canPlaceItemOn(@org.jetbrains.annotations.NotNull ItemStack item, @org.jetbrains.annotations.Nullable gg.projecteden.parchment.OptionalHumanEntity player, @org.jetbrains.annotations.NotNull org.bukkit.block.Block block, @org.jetbrains.annotations.NotNull org.bukkit.block.BlockFace face);
+    // Parchment end
 }
