From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Fri, 18 Jun 2021 03:18:47 -0400
Subject: [PATCH] Add UnsafeValues#canPlaceItemOn


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 11bccf6c27b8b2a2ed558a1375a3b3adc7952076..f8071399ad218e30d06de5835df1cef1d77dc2d8 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -279,4 +279,18 @@ public interface UnsafeValues {
     <A extends Keyed, M> io.papermc.paper.registry.tag.@Nullable Tag<A> getTag(io.papermc.paper.registry.tag.@NotNull TagKey<A> tagKey); // Paper - hack to get tags for non-server backed registries
 
     ItemStack createEmptyStack(); // Paper - proxy ItemStack
+
+    // Parchment start
+    /**
+     * Determines if this item can be placed on a block. Must be run synchronously.
+     * Factors in player location, blocking events, hitboxes, and more.
+     * @param item item to test
+     * @param player optional player placing the block
+     * @param block block that is being placed against
+     * @param face cardinal direction
+     * @return if the item can be placed
+     */
+    java.util.concurrent.CompletableFuture<Boolean> canPlaceItemOn(@org.jetbrains.annotations.NotNull ItemStack item, @org.jetbrains.annotations.Nullable gg.projecteden.parchment.OptionalHumanEntity player, @org.jetbrains.annotations.NotNull org.bukkit.block.Block block, @org.jetbrains.annotations.NotNull org.bukkit.block.BlockFace face);
+    // Parchment end
+
 }
