From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 18:47:01 -0400
Subject: [PATCH] Expose NMS ordinals


diff --git a/src/main/java/org/bukkit/UnsafeValues.java b/src/main/java/org/bukkit/UnsafeValues.java
index 107f356026397f1a6ef8580c7f239a02bf8105fe..12b9d9fff1313aa27a8178456934b88df76580d6 100644
--- a/src/main/java/org/bukkit/UnsafeValues.java
+++ b/src/main/java/org/bukkit/UnsafeValues.java
@@ -250,5 +250,12 @@ public interface UnsafeValues {
      * @return if the item can be placed
      */
     java.util.concurrent.CompletableFuture<Boolean> canPlaceItemOn(@org.jetbrains.annotations.NotNull ItemStack item, @org.jetbrains.annotations.Nullable gg.projecteden.parchment.OptionalHumanEntity player, @org.jetbrains.annotations.NotNull org.bukkit.block.Block block, @org.jetbrains.annotations.NotNull org.bukkit.block.BlockFace face);
+
+    /**
+     * Gets the ID/ordinal of an entity according to NMS.
+     * @param type entity to get
+     * @return entity's ID
+     */
+    int entityID(org.bukkit.entity.EntityType type);
     // Parchment end
 }
