From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 68206cf0178c26c0f528a1e14a5fb4e9ad410369..24fa398cc02f803fbb3222082130bf0b7bdd391f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2053,6 +2053,22 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     UnsafeValues getUnsafe();
 
+    // Parchment start
+    /**
+     * Gets the executor which runs tasks on an asynchronous thread pool.
+     * @return async executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getAsyncExecutor();
+
+    /**
+     * Gets the executor which runs tasks on the main thread.
+     * @return main executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getMainExecutor();
+    // Parchment end
+
     // Spigot start
     public class Spigot {
 
