From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2204336d8800311b65e894739ab1b27273e7c6f2..2c35578c85be5f47c2b2a11f1df16db51e31cc41 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2139,4 +2139,20 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+
+    // Parchment start
+    /**
+     * Gets the executor which runs tasks on an asynchronous thread pool.
+     * @return async executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getAsyncExecutor();
+
+    /**
+     * Gets the executor which runs tasks on the main thread.
+     * @return main executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getMainExecutor();
+    // Parchment end
 }
