From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index cca362e54d6ff4a5a1e60f85a7eb1b3d222d3d48..e9ee90ce4f1c8b58ab2b20a6806de11dcdf98423 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2003,4 +2003,20 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+
+    // Parchment start
+    /**
+     * Gets the executor which runs tasks on an asynchronous thread pool.
+     * @return async executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getAsyncExecutor();
+
+    /**
+     * Gets the executor which runs tasks on the main thread.
+     * @return main executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getMainExecutor();
+    // Parchment end
 }
