From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e48af3822e9f118399c3a1c9358c56efae12e0da..1689306bb9a9fc251574877427ca846e6def7003 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1888,4 +1888,20 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
+
+    // Parchment start
+    /**
+     * Gets the executor which runs tasks on an asynchronous thread pool.
+     * @return async executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getAsyncExecutor();
+
+    /**
+     * Gets the executor which runs tasks on the main thread.
+     * @return main executor
+     */
+    @NotNull
+    java.util.concurrent.Executor getMainExecutor();
+    // Parchment end
 }
