From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index aa99410a0bee71e68da90b596134632f3a6364fe..4390fa4cc4616a13855d38cb5e09d60153368428 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1851,6 +1851,17 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @ApiStatus.Experimental
     public boolean canSee(@NotNull Entity entity);
 
+    // Parchment start
+    /**
+     * Gets a view of the hidden entity UUIDs.
+     *
+     * @param plugin Plugin that has hidden entities
+     * @return a view of hidden entity UUIDs
+     */
+    @ApiStatus.Experimental
+    public java.util.@NotNull Set<java.util.UUID> getHiddenEntities(@NotNull Plugin plugin);
+    // Parchment end
+
     /**
      * Checks to see if this player is currently flying or not.
      *
