From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 53957259b284a7cfbfcbb6d64c5a981999d161d6..1c42b484882a5e6f2c8276c428824b2cbf6a39e3 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1433,6 +1433,17 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @org.jetbrains.annotations.ApiStatus.Experimental // Paper
     public boolean canSee(@NotNull Entity entity);
 
+    // Parchment start
+    /**
+     * Gets a view of the hidden entity UUIDs.
+     *
+     * @param plugin Plugin that has hidden entities
+     * @return a view of hidden entity UUIDs
+     */
+    @org.jetbrains.annotations.ApiStatus.Experimental
+    public java.util.@NotNull Set<UUID> getHiddenEntities(@NotNull Plugin plugin);
+    // Parchment end
+
     /**
      * Checks to see if this player is currently flying or not.
      *
