From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 558903782a1898d875274be7561f332f1ca568b2..c8f6fb1d64c86ba2ca89793e25610d1a88f78540 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1951,6 +1951,17 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @ApiStatus.Experimental
     public boolean canSee(@NotNull Entity entity);
 
+    // Parchment start
+    /**
+     * Gets a view of the hidden entity UUIDs.
+     *
+     * @param plugin Plugin that has hidden entities
+     * @return a view of hidden entity UUIDs
+     */
+    @ApiStatus.Experimental
+    public java.util.@NotNull Set<java.util.UUID> getHiddenEntities(@NotNull Plugin plugin);
+    // Parchment end
+
     // Paper start
     /**
      * Returns whether the {@code other} player is listed for {@code this}.
