From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 42e186ffdb421ccea61c4325f4b9800d55f624a8..8b2380241dbf40bd03941250a66e0a7b4a2282ec 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2092,6 +2092,17 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public boolean canSee(@NotNull Entity entity);
 
+    // Parchment start
+    /**
+     * Gets a view of the hidden entity UUIDs.
+     *
+     * @param plugin Plugin that has hidden entities
+     * @return a view of hidden entity UUIDs
+     */
+    public java.util.@NotNull Set<java.util.UUID> getHiddenEntities(@NotNull Plugin plugin);
+    // Parchment end
+
+
     // Paper start
     /**
      * Returns whether the {@code other} player is listed for {@code this}.
