From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Wed, 15 Mar 2023 01:05:09 -0400
Subject: [PATCH] Revert BungeeChat Deprecation


diff --git a/build.gradle.kts b/build.gradle.kts
index 7dab3328c111b9d4e06a37f25f03708870f4bab9..eefb9b19fb76689a5e7ba86a17ef7117ad67789b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -48,7 +48,7 @@ dependencies {
     api("com.google.guava:guava:32.1.2-jre")
     api("com.google.code.gson:gson:2.10.1")
     // Paper start - adventure
-    api("net.md-5:bungeecord-chat:$bungeeCordChatVersion-deprecated+build.18") {
+    api("net.md-5:bungeecord-chat:$bungeeCordChatVersion") { // Parchment - revert
         exclude("com.google.guava", "guava")
     }
     // Paper - adventure
@@ -172,7 +172,7 @@ tasks.withType<Javadoc> {
         "https://guava.dev/releases/32.1.2-jre/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/",
         "https://javadoc.io/doc/org.jetbrains/annotations/$annotationsVersion/", // Paper - we don't want Java 5 annotations
-        // "https://javadoc.io/doc/net.md-5/bungeecord-chat/$bungeeCordChatVersion/", // Paper - don't link to bungee chat
+        "https://javadoc.io/doc/net.md-5/bungeecord-chat/$bungeeCordChatVersion/", // Paper - don't link to bungee chat
         // Paper start - add missing javadoc links
         "https://javadoc.io/doc/org.joml/joml/1.10.5/index.html",
         "https://www.javadoc.io/doc/com.google.code.gson/gson/2.10.1",
