From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Wed, 15 Mar 2023 01:05:09 -0400
Subject: [PATCH] Revert BungeeChat Deprecation


diff --git a/build.gradle.kts b/build.gradle.kts
index 3c4dd6ebc2289c44c2f5723e7920aadffdc51884..1d7d810de20536fee67bf6073aab3ff24f01c55d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -25,7 +25,7 @@ dependencies {
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:31.1-jre")
     api("com.google.code.gson:gson:2.10")
-    api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.6") // Paper
+    api("net.md-5:bungeecord-chat:1.16-R0.4") // Parchment
     api("org.yaml:snakeyaml:1.33")
     // Paper start
     api("com.googlecode.json-simple:json-simple:1.1.1") {
@@ -104,7 +104,7 @@ tasks.withType<Javadoc> {
         "https://javadoc.io/doc/org.yaml/snakeyaml/1.33/",
         "https://javadoc.io/doc/org.jetbrains/annotations/23.0.0/", // Paper - we don't want Java 5 annotations
         // Paper start
-        //"https://javadoc.io/doc/net.md-5/bungeecord-chat/1.16-R0.4/", // don't link to bungee chat
+        "https://javadoc.io/doc/net.md-5/bungeecord-chat/1.16-R0.4/", // don't link to bungee chat
         "https://jd.adventure.kyori.net/api/$adventureVersion/",
         "https://jd.adventure.kyori.net/text-minimessage/$adventureVersion/",
         "https://jd.adventure.kyori.net/text-serializer-gson/$adventureVersion/",
