From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sat, 19 Jun 2021 18:03:59 -0400
Subject: [PATCH] Temporary Access Transformers


diff --git a/src/main/java/net/minecraft/world/item/BlockItem.java b/src/main/java/net/minecraft/world/item/BlockItem.java
index b0204af850ee182773ad458208cccd946ad148d5..92980f04bc1bac9924907007eb63fd04eba3875c 100644
--- a/src/main/java/net/minecraft/world/item/BlockItem.java
+++ b/src/main/java/net/minecraft/world/item/BlockItem.java
@@ -207,7 +207,7 @@ public class BlockItem extends Item {
         }).orElse(state);
     }
 
-    protected boolean canPlace(BlockPlaceContext context, BlockState state) {
+    public boolean canPlace(BlockPlaceContext context, BlockState state) { // Parchment
         Player entityhuman = context.getPlayer();
         CollisionContext voxelshapecollision = entityhuman == null ? CollisionContext.empty() : CollisionContext.of(entityhuman);
         // CraftBukkit start - store default return
