From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:09:16 -0400
Subject: [PATCH] Add Player#setGameProfile


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 66866e1dabe79a78df255fc17842bdb36988d86f..fd0c584d459f912a714e464e40803d4e3dca8bd8 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1510,6 +1510,12 @@ public abstract class Player extends LivingEntity {
         return this.gameProfile;
     }
 
+    // Parchment start
+    public void setProfile(GameProfile gameProfile) {
+        this.gameProfile = gameProfile;
+    }
+    // Parchment end
+
     public Inventory getInventory() {
         return this.inventory;
     }
