From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f2ed626ffb125d5a8663f2021524e5359b203177..e9de3ee7fbc1b9ff59802f6823312e974bfcc33c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3227,4 +3227,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Parchment start
+    @Override
+    public java.util.concurrent.Executor getAsyncExecutor() {
+        return io.papermc.paper.util.MCUtil.asyncExecutor;
+    }
+
+    @Override
+    public java.util.concurrent.Executor getMainExecutor() {
+        return io.papermc.paper.util.MCUtil.MAIN_EXECUTOR;
+    }
+    // Parchment end
 }
