From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index ce3bbb9909dde2594d33fd33d5d93a5afbbe601f..e6186484a966431b562c75f3f0952637f30ed221 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3229,4 +3229,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Parchment start
+    @Override
+    public java.util.concurrent.Executor getAsyncExecutor() {
+        return io.papermc.paper.util.MCUtil.asyncExecutor;
+    }
+
+    @Override
+    public java.util.concurrent.Executor getMainExecutor() {
+        return io.papermc.paper.util.MCUtil.MAIN_EXECUTOR;
+    }
+    // Parchment end
 }
