From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:09:16 -0400
Subject: [PATCH] Add Player#setGameProfile


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index bc6bdc8752718777ac698cfda2a0e7f35fcf9896..05584d344c9f7d6d89bb8fbdfa254c1dca91b590 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1511,6 +1511,12 @@ public abstract class Player extends LivingEntity {
         return this.gameProfile;
     }
 
+    // Parchment start | TODO: rename to match getter
+    public void setProfile(GameProfile gameProfile) {
+        this.gameProfile = gameProfile;
+    }
+    // Parchment end
+
     @Nullable
     public ProfilePublicKey getProfilePublicKey() {
         return this.profilePublicKey;
