From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:09:16 -0400
Subject: [PATCH] Add Player#setGameProfile


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 07f8e2bfabd9292b9f594bf408d40fc939497a1c..6237d023fe6e449077cb53d93f003ae92dc6b4a0 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1504,6 +1504,12 @@ public abstract class Player extends LivingEntity {
         return this.gameProfile;
     }
 
+    // Parchment start
+    public void setProfile(GameProfile gameProfile) {
+        this.gameProfile = gameProfile;
+    }
+    // Parchment end
+
     public Inventory getInventory() {
         return this.inventory;
     }
