From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index cbf96eeaed245632e5386f824168ba42113038e1..d01354f10d0db28fc5711cf4d598cdc01fa4cc80 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2667,4 +2667,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Parchment start
+    @Override
+    public java.util.concurrent.Executor getAsyncExecutor() {
+        return net.minecraft.server.MCUtil.asyncExecutor;
+    }
+
+    @Override
+    public java.util.concurrent.Executor getMainExecutor() {
+        return net.minecraft.server.MCUtil.MAIN_EXECUTOR;
+    }
+    // Parchment end
 }
