From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3336f223f9f0c6207471de15326f40d2ddc1dae2..a193f2ab16fecc26c04408a4473e064dd79168dd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2833,4 +2833,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Parchment start
+    @Override
+    public java.util.concurrent.Executor getAsyncExecutor() {
+        return net.minecraft.server.MCUtil.asyncExecutor;
+    }
+
+    @Override
+    public java.util.concurrent.Executor getMainExecutor() {
+        return net.minecraft.server.MCUtil.MAIN_EXECUTOR;
+    }
+    // Parchment end
 }
