From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Sun, 20 Jun 2021 19:39:38 -0400
Subject: [PATCH] Expose MCUtil Executors


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f4847a8671bd21f78af598faacea5c763669ddbc..97e60454716eacdfa4420efff2596b7f99d2bfd6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2869,4 +2869,16 @@ public final class CraftServer implements Server {
     }
 
     // Paper end
+
+    // Parchment start
+    @Override
+    public java.util.concurrent.Executor getAsyncExecutor() {
+        return net.minecraft.server.MCUtil.asyncExecutor;
+    }
+
+    @Override
+    public java.util.concurrent.Executor getMainExecutor() {
+        return net.minecraft.server.MCUtil.MAIN_EXECUTOR;
+    }
+    // Parchment end
 }
