From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lexikiq <noellekiq@gmail.com>
Date: Mon, 13 Sep 2021 22:28:59 -0400
Subject: [PATCH] Disable sleep status announcements


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index b012ef12b6ec89ed8281d8b475c87f53fee196c2..8079f25e2830b73b48d32b2abecf6052c3005dea 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1046,7 +1046,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     }
 
     private void announceSleepStatus() {
-        if (this.canSleepThroughNights()) {
+        if (false && this.canSleepThroughNights()) {
             if (!this.getServer().isSingleplayer() || this.getServer().isPublished()) {
                 int i = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
                 MutableComponent ichatmutablecomponent;
