From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 14:41:24 -0400
Subject: [PATCH] Add HangingFrame Tick API


diff --git a/src/main/java/net/minecraft/world/entity/decoration/HangingEntity.java b/src/main/java/net/minecraft/world/entity/decoration/HangingEntity.java
index f88918b7c15b72f4cb1dd313734f550432fd5858..74215934d25d48fc2628401ed57cb78222b8c2ce 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -41,6 +41,7 @@ public abstract class HangingEntity extends Entity {
     private int checkInterval; { this.checkInterval = this.getId() % this.level().spigotConfig.hangingTickFrequency; } // Paper - Perf: offset item frame ticking
     public BlockPos pos;
     protected Direction direction;
+    public boolean tick = true; // Parchment
 
     protected HangingEntity(EntityType<? extends HangingEntity> type, Level world) {
         super(type, world);
@@ -116,7 +117,7 @@ public abstract class HangingEntity extends Entity {
 
     @Override
     public void tick() {
-        if (!this.level().isClientSide) {
+        if (tick && !this.level().isClientSide) { // Parchment
             this.checkBelowWorld();
             if (this.checkInterval++ == this.level().spigotConfig.hangingTickFrequency) { // Spigot
                 this.checkInterval = 0;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHanging.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHanging.java
index cbdcf767c01e2c173913f56747d2dacbda2f0094..a74f783df0225223a2b2ae3dc632541ccf078893 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHanging.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHanging.java
@@ -66,4 +66,16 @@ public class CraftHanging extends CraftEntity implements Hanging {
     public String toString() {
         return "CraftHanging";
     }
+
+    // Parchment start
+    @Override
+    public boolean canTick() {
+        return this.getHandle().tick;
+    }
+
+    @Override
+    public void setCanTick(boolean tick) {
+        this.getHandle().tick = tick;
+    }
+    // Parchment end
 }
