From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 7e6116963d835d4606ef3d93b69d3e44b61288e1..1da9415f141aabe8df5b82392c142c1f3fd960fb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2145,6 +2145,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.server.getPluginManager().callEvent(new PlayerShowEntityEvent(this, entity));
     }
     // Paper start
+
+    // Parchment start
+    @Override
+    public Set<UUID> getHiddenEntities(Plugin plugin) {
+        return invertedVisibilityEntities.entrySet().stream()
+            .filter(entry -> entry.getValue().contains(CraftPlayer.getPluginWeakReference(plugin)))
+            .map(Map.Entry::getKey)
+            .collect(java.util.stream.Collectors.toSet());
+    }
+   // Parchment end
+
     @Override
     public void setPlayerProfile(com.destroystokyo.paper.profile.PlayerProfile profile) {
         ServerPlayer self = this.getHandle();
