From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index dbdeb913e228651cadf5dbd7ec98afc738c80522..64862c4bd3ec5a1ca45da6cd491c7935b850947c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1940,6 +1940,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         server.getPluginManager().callEvent(new PlayerShowEntityEvent(this, entity)); // Paper
     }
     // Paper start
+
+    // Parchment start
+    @Override
+    public Set<UUID> getHiddenEntities(Plugin plugin) {
+        return invertedVisibilityEntities.entrySet().stream()
+            .filter(entry -> entry.getValue().contains(CraftPlayer.getPluginWeakReference(plugin)))
+            .map(Map.Entry::getKey)
+            .collect(java.util.stream.Collectors.toSet());
+    }
+   // Parchment end
+
     @Override
     public void setPlayerProfile(com.destroystokyo.paper.profile.PlayerProfile profile) {
         ServerPlayer self = this.getHandle();
