From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index d7ba155d9485128f4c7ad8ab4a2b7302a86aeb67..16af80579093c5e7f8c8183d36f606a5d13cbd93 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1766,6 +1766,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             registerEntity(player);
         }
     }
+
+    // Parchment start
+    @Override
+    public Set<UUID> getHiddenEntities(Plugin plugin) {
+        return hiddenEntities.entrySet().stream()
+            .filter(entry -> entry.getValue().contains(CraftPlayer.getPluginWeakReference(plugin)))
+            .map(Map.Entry::getKey)
+            .collect(java.util.stream.Collectors.toSet());
+    }
+    // Parchment end
+
     public void setPlayerProfile(com.destroystokyo.paper.profile.PlayerProfile profile) {
         ServerPlayer self = getHandle();
         self.gameProfile = com.destroystokyo.paper.profile.CraftPlayerProfile.asAuthlibCopy(profile);
