From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lexi <lexi@qixils.dev>
Date: Thu, 23 Jun 2022 22:03:18 -0400
Subject: [PATCH] Add Player#getHiddenEntities API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a5220a4e514865c70700527bcd2f941d852564c0..0950b5f97b3382d1e41da9837e1e0b7cce6ba1b4 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1838,6 +1838,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             registerEntity(player);
         }
     }
+
+    // Parchment start
+    @Override
+    public Set<UUID> getHiddenEntities(Plugin plugin) {
+        return hiddenEntities.entrySet().stream()
+            .filter(entry -> entry.getValue().contains(CraftPlayer.getPluginWeakReference(plugin)))
+            .map(Map.Entry::getKey)
+            .collect(java.util.stream.Collectors.toSet());
+    }
+    // Parchment end
+
     public void setPlayerProfile(com.destroystokyo.paper.profile.PlayerProfile profile) {
         ServerPlayer self = getHandle();
         self.gameProfile = com.destroystokyo.paper.profile.CraftPlayerProfile.asAuthlibCopy(profile);
