From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Wed, 20 Jul 2022 05:26:25 -0400
Subject: [PATCH] Add Timings Events


diff --git a/src/main/java/co/aikar/timings/TimingsExport.java b/src/main/java/co/aikar/timings/TimingsExport.java
index 06bff37e4c1fddd3be6343049a66787c63fb420c..b6567f558cd25698208a06513c241646e0b28341 100644
--- a/src/main/java/co/aikar/timings/TimingsExport.java
+++ b/src/main/java/co/aikar/timings/TimingsExport.java
@@ -349,6 +349,8 @@ public class TimingsExport extends Thread {
             timingsURL = con.getHeaderField("Location");
             listeners.sendMessage(text("View Timings Report: ", NamedTextColor.GREEN).append(text(timingsURL).clickEvent(ClickEvent.clickEvent(ClickEvent.Action.OPEN_URL, timingsURL))));
 
+            new co.aikar.timings.event.TimingsGenerateReportEvent(this.listeners, timingsURL).callEvent();
+
             if (response != null && !response.isEmpty()) {
                 Bukkit.getLogger().log(Level.INFO, "Timing Response: " + response);
             }
