From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Thu, 25 Aug 2022 20:09:40 -0400
Subject: [PATCH] Disable set respawn message


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 071e38a48114b6cb0155c1dbd60628cdc81189e8..0297c95ca67f5f113b696b4cbd149119b92eca2f 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2523,9 +2523,10 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
             forced = event.isForced();
             // Paper end - Add PlayerSetSpawnEvent
 
-            if (event.willNotifyPlayer() && event.getNotification() != null) { // Paper - Add PlayerSetSpawnEvent
-                this.sendSystemMessage(PaperAdventure.asVanilla(event.getNotification())); // Paper - Add PlayerSetSpawnEvent
-            }
+            /* Parchment start
+            if (event.willNotifyPlayer() && event.getNotification() != null) { // Paper
+                this.sendSystemMessage(PaperAdventure.asVanilla(event.getNotification())); // Paper
+            } Parchment end */
 
             this.respawnPosition = pos;
             this.respawnDimension = dimension;
