From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Thu, 25 Aug 2022 20:09:40 -0400
Subject: [PATCH] Disable set respawn message


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 823bc659f7421c3752e4cf1333e1568e1308464d..b7cdefdbbdba46fdd9f39275474be356c2af8b40 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2399,9 +2399,11 @@ public class ServerPlayer extends Player {
             forced = event.isForced();
             // Paper end
 
+            /* Parchment start
             if (event.willNotifyPlayer() && event.getNotification() != null) { // Paper
                 this.sendSystemMessage(PaperAdventure.asVanilla(event.getNotification())); // Paper
             }
+            Parchment end */
 
             this.respawnPosition = pos;
             this.respawnDimension = dimension;
