From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Thu, 25 Aug 2022 20:09:40 -0400
Subject: [PATCH] Disable set respawn message


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 2abe06cb5de5ac63c31c1a120a3239aca07557f8..2eae1123cc24ad49248a8668701aa7015cbc4840 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2279,9 +2279,11 @@ public class ServerPlayer extends Player {
             forced = event.isForced();
             // Paper end
 
+            /* Parchment start
             if (event.willNotifyPlayer() && event.getNotification() != null) { // Paper
                 this.sendSystemMessage(PaperAdventure.asVanilla(event.getNotification())); // Paper
             }
+            Parchment end */
 
             this.respawnPosition = pos;
             this.respawnDimension = dimension;
