From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-Mc <cjblanton2@gmail.com>
Date: Mon, 9 Jan 2023 19:21:16 -0500
Subject: [PATCH] Remove Unneeded Override


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 1bf7e21cd3b981013ff634dfc070d572af491da9..a45de4db8a73cea947c76771f9e540f6e64763bb 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1737,7 +1737,6 @@ public class ServerLevel extends Level implements WorldGenLevel {
         // Parchment end
     }
 
-    @Override
     public void playSeededSound(@org.jetbrains.annotations.Nullable Player except, Entity entity, SoundEvent sound, SoundSource category, float volume, float pitch, long seed) {
         // Parchment start - sound event
         CraftEventFactory.playSoundEvent(new gg.projecteden.parchment.event.sound.SoundEvent(
