From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Thu, 23 Mar 2023 18:18:43 -0400
Subject: [PATCH] Make SynchedEntityData#packAll Public


diff --git a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
index 15add3f4dfd718ec09bb1db4f22223466936879c..ab6d67d7acf657982bbd34af8a60e1bb36168979 100644
--- a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
@@ -267,7 +267,7 @@ public class SynchedEntityData {
     // We need to pack all as we cannot rely on "non default values" or "dirty" ones.
     // Because these values can possibly be desynced on the client.
     @Nullable
-    private List<SynchedEntityData.DataValue<?>> packAll() {
+    public List<SynchedEntityData.DataValue<?>> packAll() {
         if (this.isEmpty()) {
             return null;
         }
