From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Thu, 23 Mar 2023 18:18:43 -0400
Subject: [PATCH] Make SynchedEntityData#packAll Public


diff --git a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
index 5dfb35117c285e0b202dc9c088ad5848beb8d054..662e20a607c348403250f50a8ff0a0171e128701 100644
--- a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
@@ -289,7 +289,7 @@ public class SynchedEntityData {
     // We need to pack all as we cannot rely on "non default values" or "dirty" ones.
     // Because these values can possibly be desynced on the client.
     @Nullable
-    private List<SynchedEntityData.DataValue<?>> packAll() {
+    public List<SynchedEntityData.DataValue<?>> packAll() {
         if (this.isEmpty()) {
             return null;
         }
