From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Blast-MC <cjblanton2@gmail.com>
Date: Sat, 4 Jan 2025 19:16:06 -0500
Subject: [PATCH] Nether Portal Kick Fix


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 544882862b2248d426185dd67bd217f593caef2a..0474a5d6a35955702ba76bdc24f5c6e77e358a90 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1793,7 +1793,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         if (this.isAffectedByBlocks()) {
             AABB axisalignedbb = this.getBoundingBox().deflate(9.999999747378752E-6D);
             LongSet longset = this.visitedBlocks;
-            Iterator iterator = queuedCollisionChecks.iterator();
+            Iterator iterator = new ArrayList<>(queuedCollisionChecks).iterator();
 
             while (iterator.hasNext()) {
                 Entity.Movement entity_b = (Entity.Movement) iterator.next();
