--- a/src/main/java/org/bukkit/event/world/GenericGameEvent.java
+++ b/src/main/java/org/bukkit/event/world/GenericGameEvent.java
@@ -15,7 +_,7 @@
  * Specific Bukkit events should be used where possible, this event is mainly
  * used internally by Sculk sensors.
  */
-public class GenericGameEvent extends WorldEvent implements Cancellable {
+public class GenericGameEvent extends WorldEvent implements Cancellable, gg.projecteden.parchment.HasLocation { // Parchment
 
     private static final HandlerList handlers = new HandlerList();
     private final GameEvent event;
