--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -59,7 +_,7 @@
     }
     // CraftBukkit end
     private static final Component DEFAULT_NAME = Component.translatable("container.barrel");
-    private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
+    private NonNullList<ItemStack> items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
     public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
         // Paper start - delay open/close callbacks
         @Override
@@ -118,7 +_,7 @@
 
     @Override
     public int getContainerSize() {
-        return 27;
+        return 54;
     }
 
     @Override
@@ -138,7 +_,7 @@
 
     @Override
     protected AbstractContainerMenu createMenu(final int containerId, final Inventory inventory) {
-        return ChestMenu.threeRows(containerId, inventory, this);
+        return ChestMenu.sixRows(containerId, inventory, this);
     }
 
     @Override
